metadata/metadata.go

```diff
@@ -222,7 +249,7 @@ func FromOutgoingContext(ctx context.Context) (MD, bool) {
 		return nil, false
 	}
 
-	out := MD{}
+	out := make(MD, len(raw.md)+len(raw.added))
```

Comment:
Technically this should be `len(raw.md) + sum(lengths of raw.added sub-slices)/2`.  (And that's a worst case; there could be duplicate keys in `raw.added` as well.)  I'm not certain it's worth it to pre-scan to compute, but if it avoids a map reallocation, it probably is.  It's unlikely many people are using `Append` anyway, so for most cases `raw.added` will be `nil` (AKA precomputing is usually ~free).

Reply:
so, should I add size precomputing? I think it's a bit excessively.

Reply:
Yes. The cost of a reallocation is very large, and this change means we could be changing from the default map size to a smaller size, increasing the risk of needing to reallocate and making this perform worse.  We should instead pre-compute a worst case for pre-allocation, or just stick with the default resizing behavior.